### PR TITLE
squashing last four commits

### PR DESCRIPTION
**Pull request by adhaka**
The pull request contains:

a) RMSPRop, Adagrad, ADAM optimizers with MCSE and EKBO stopping rules.
b) Convergence and other diagnostics 
c) Notebook on robust regression using all optimizers ..

### CODE REVIEW

File: viabel/optimization_diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
+
+import autograd.numpy as np
+from autograd.extend import primitive
+
+def compute_R_hat(chains, warmup=500):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:5]: Please add function description (as in, e.g., montecarlo_se)

File: viabel/optimization_diagnostics.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    psi_dot_dot = np.mean(psi_dot_j, axis=0)
+    s_j_2 = np.sum((psi - np.expand_dims(psi_dot_j, axis=1)) ** 2, axis=1) / (n_iters - 1)
+    B = n_iters * np.sum((psi_dot_j - psi_dot_dot) ** 2, axis=0) / (n_chains2 - 1)
+    W = np.nanmean(s_j_2, axis=0)
+    W = W + jitter
+    var_hat = (n_iters - 1) / n_iters + (B / (n_iters*W))
+    R_hat = np.sqrt(var_hat)
+    return var_hat, R_hat
+
+def compute_R_hat_halfway(chains, interval=100, start=1000):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:30]: Please add function description (as in, e.g., montecarlo_se2)

File: viabel/optimization_diagnostics.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    for i in range(n_subchains):
+        sub_chains = chains[:, :start+(i+1)*interval,:]
+        n_sub_chains, n_sub_iters, K = sub_chains.shape
+        r_hat_current = compute_R_hat(sub_chains, warmup=n_sub_iters//2)[1]
+        r_hats_halfway.append(r_hat_current)
+
+    return np.array(r_hats_halfway)
+
+def compute_R_hat_halfway_light(chains, warmup_fraction=0.5):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:43]: Please add function description (as in, e.g., montecarlo_se2)

File: viabel/optimization_diagnostics.py (hunk 1 above)
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:5]: Please add function description (as in, e.g., montecarlo_se2)

File: viabel/optimization_diagnostics.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        mean_var_chains = np.mean(variances, axis=0)
+        mcse_per_chain = np.sqrt(variances / n_iters)
+        mcse_combined = np.sqrt(variances_combined/(i*n_chains))
+        mcse_combined_list[i] = mcse_combined
+
+    #print(mcse_per_chain.shape)
+    return mcse_per_chain, mcse_combined_list
+
+# compute mcmcse for a chain/array
+def monte_carlo_se2(iterate_chains, warmup=500, param_idx=0):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:69]: How is this different from monte_carlo_se2? Please use more descriptive names to distinguish these

File: viabel/optimization_diagnostics.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        if increasing:
+            filtered_chain = filtered_chain -np.min(filtered_chain)
+        else:
+            filtered_chain = filtered_chain -np.min(filtered_chain)
+        k_post, _ = gpdfit(filtered_chain)
+        k_hat_values[i] = k_post
+
+    return np.nanmax(k_hat_values)
+
+def compute_posterior_moments(prior_mean, prior_covariance, noise_variance, x, y):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:192]: Please add function description (as in, e.g., montecarlo_se2)

File: viabel/optimization_diagnostics.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    S_precision = prior_precision + x.T @ x *(1. / noise_variance)
+    a = np.linalg.cholesky(S_precision)
+    tmp1 = np.linalg.inv(a)
+    S = tmp1.T @ tmp1
+    post_S=S
+    post_mu = prior_precision@prior_mean + (1./noise_variance)* x.T@ y
+    post_mu = post_S@ post_mu
+    return post_mu, post_S
+
+def get_samples_and_log_weights(logdensity, var_family, var_param, n_samples):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:205]: Please add function description (as in, e.g., montecarlo_se2)

File: viabel/optimizers.py [hunk 7]
```diff
@@ -0,0 +1,808 @@
...
+import matplotlib.pyplot as plt
+
+import tqdm
+import scipy.stats as stats
+from  .optimization_diagnostics import autocorrelation, monte_carlo_se, monte_carlo_se2, compute_khat_iterates, \
+    gpdfit, montecarlo_se, compute_R_hat, compute_R_hat_halfway, compute_R_hat_halfway_light
+
+
+__all__ = [
+    'adagrad_workflow_optimize',
```
Comment (jhuggins) [inline viabel/optimizers.py:17]: What's the difference between adagrad_workflow_optimize and adagrad_optimize?

File: viabel/optimizers.py [hunk 8]
```diff
@@ -0,0 +1,808 @@
...
+from functools import partial
+import matplotlib.pyplot as plt
+
+import tqdm
+import scipy.stats as stats
+from  .optimization_diagnostics import autocorrelation, monte_carlo_se, monte_carlo_se2, compute_khat_iterates, \
+    gpdfit, montecarlo_se, compute_R_hat, compute_R_hat_halfway, compute_R_hat_halfway_light
+
+
+__all__ = [
```
Comment (jhuggins) [inline viabel/optimizers.py:16]: Please move adagrad_optimize to this file and update other references in the code and examples as necessary

File: viabel/optimizers.py [hunk 9]
```diff
@@ -0,0 +1,808 @@
...
+
+    if stopping_rule ==1:
+        smoothed_opt_param = np.mean(variational_param_history_array[start_stats:,:], axis=0)
+    elif stopping_rule ==2 and sto_process_convergence== True:
+        smoothed_opt_param = np.mean(variational_param_post_conv_history_chains[0, :], axis=0)
+
+    if stopping_rule == 2 and sto_process_convergence == False:
+        smoothed_opt_param = np.mean(variational_param_history_array[start_stats:,:], axis=0)
+
+    if plotting:
```
Comment (jhuggins) [inline viabel/optimizers.py:223]: Plotting functionality should be separated out

File: viabel/optimizers.py [hunk 10]
```diff
@@ -0,0 +1,808 @@
...
+    gpdfit, montecarlo_se, compute_R_hat, compute_R_hat_halfway, compute_R_hat_halfway_light
+
+
+__all__ = [
+    'adagrad_workflow_optimize',
+    'rmsprop_workflow_optimize',
+    'adam_workflow_optimize'
+]
+
+def learning_rate_schedule(n_iters, learning_rate, learning_rate_end):
```
Comment (jhuggins) [inline viabel/optimizers.py:22]: Add description for this function

File: viabel/optimizers.py [hunk 11]
```diff
@@ -0,0 +1,808 @@
...
+        end_decrease_at = 3*n_iters//4
+    for i in range(n_iters):
+        if learning_rate_end is None or i < start_decrease_at:
+            yield learning_rate
+        elif i < end_decrease_at:
+            yield a / (b + i - start_decrease_at + 1)
+        else:
+            yield learning_rate_end
+
+def adagrad_workflow_optimize(n_iters, objective_and_grad, init_param,K,
```
Comment (jhuggins) [inline viabel/optimizers.py:42]: Please add some additional comments to the optimize functions that explain what each part is doing. Perhaps it's possible to break these functions into smaller pieces  to make the code more self-documenting?

**Review by jhuggins — CHANGES_REQUESTED** [review]

File: viabel/optimization_diagnostics.py [hunk 12]
```diff
@@ -40,45 +72,56 @@ def compute_R_hat_halfway(chains, interval=100, start=1000):
...
 
 # compute mcmcse for a chain/array
-def monte_carlo_se2(iterate_chains, warmup=500, param_idx=0):
-    '''
-    compute monte carlo standard error for all chains and one parameter at a time.
-    :param iterate_chains:
-    :param warmup: warmup iterates
-    :return: array of mcse error for all chains for a particular variational parameter.
-    '''
+def monte_carlo_se_moving(iterate_chains, warmup=500, param_idx=0):
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:97]: in this function and other below, make warmup default to half of chain length or something similar. hardcoded value is not robust

File: viabel/optimization_diagnostics.py [hunk 13]
```diff
@@ -142,7 +207,7 @@ def autocorrelation(iterate_chains, warmup=500, param_idx=0, lag_max=80):
     lag = 1
     a = []
     neff_array = []
-    while lag < 100:
+    while lag < lag_max:
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:210]: use for loop over range

File: viabel/optimization_diagnostics.py [hunk 14]
```diff
@@ -156,16 +221,25 @@ def autocorrelation(iterate_chains, warmup=500, param_idx=0, lag_max=80):
...
+    ----------
+    iterate_chains : multi-dimensional array, shape=(n_chains, n_iters, n_var_params)
+
+    warmup : warmup iterates
+
+    param_idx : index of the variational parameter
+
+    increasing : boolean sort array in increasing order: TRUE or decreasing order:FALSE
+
+    fraction: the fraction of iterates
```
Comment (jhuggins) [inline viabel/optimization_diagnostics.py:237]: Why have warmup and fraction? It doesn't seem to fundamentally provide any extra flexibility

**Review by jhuggins — CHANGES_REQUESTED** [review]

File: viabel/optimization_diagnostics.py (hunk 14 above)
Comment (adhaka) [reply viabel/optimization_diagnostics.py:260]: this is fixed now.

File: viabel/optimizers.py (hunk 7 above)
Comment (adhaka) [reply viabel/optimizers.py:18]: I still want to keep adagrad_optimize, so that we have one optimizer without any diagnostics somewhere in the code, this is also useful if a user wants to compare the performance of the optimizers with/without these diagnostics. But I am open to removing it if you feel so.

File: viabel/optimizers.py (hunk 9 above)
Comment (adhaka) [reply viabel/optimizers.py:334]: I will try to refactor this in the next commit ...

File: viabel/optimizers.py (hunk 10 above)
Comment (adhaka) [reply viabel/optimizers.py:24]: This is fixed now .

File: viabel/optimization_diagnostics.py [hunk 15]
```diff
@@ -27,7 +41,25 @@ def compute_R_hat(chains, warmup=500):
     R_hat = np.sqrt(var_hat)
     return var_hat, R_hat
 
-def compute_R_hat_halfway(chains, interval=100, start=1000):
+def compute_R_hat_window(chains, interval=100, start=1000):
```
[the remark this answers is not in the document]
Comment (adhaka) [reply viabel/optimization_diagnostics.py]: Yes, this is done, 100 is just a nice round value to make evaluations, but I will give another though how to improve this.

File: viabel/optimization_diagnostics.py [hunk 16]
```diff
@@ -40,45 +72,56 @@ def compute_R_hat_halfway(chains, interval=100, start=1000):
 
     return np.array(r_hats_halfway)
 
-def compute_R_hat_halfway_light(chains, warmup_fraction=0.5):
+def compute_R_hat_moving(chains, warmup_fraction=0.5):
```
[the remark this answers is not in the document]
Comment (adhaka) [reply viabel/optimization_diagnostics.py]: Thanks for the idea, this is fixed now..

File: viabel/optimization_diagnostics.py [hunk 17]
```diff
@@ -40,45 +72,56 @@ def compute_R_hat_halfway(chains, interval=100, start=1000):
...
-    :param warmup: warmup iterates
-    :return: array of mcse error for all chains for a particular variational parameter.
-    '''
+def monte_carlo_se_moving(iterate_chains, warmup=500, param_idx=0):
+    """
+    Compute the monte carlo standard error for a variational parameter
+    at each iterate using all iterates before that iterate.
+    The MCSE is computed using eq (5) of https://arxiv.org/pdf/1903.08008.pdf
+
+    Here, MCSE(\lambda_i):  sqrt(V(\lambda_i)/Seff)
```
[the remark this answers is not in the document]
Comment (adhaka) [reply viabel/optimization_diagnostics.py]: fixed.

File: viabel/optimization_diagnostics.py (hunk 12 above)
Comment (adhaka) [reply viabel/optimization_diagnostics.py]: fixed.

File: viabel/optimization_diagnostics.py (hunk 13 above)
Comment (adhaka) [reply viabel/optimization_diagnostics.py]: fixed.
